### ISSUE TITLE
Watcher does not run in MIX_ENV=prod mode

### ISSUE THREAD

**Issue by jbunce** [omisego/elixir-omg#449]
Watcher runs in `dev` mode, but the following exception is encountered when running in `prod` mode. Build `42c549a20c75a8d90db8a071bc177d6248cd2852`.

```
watcher-0 watcher Last message: :sync⋅
watcher-0 watcher 2019-02-18 04:52:56.804 [error] module=gen_server function=error_info/7 ⋅GenServer :exit_challenger terminating
watcher-0 watcher ** (stop) exited in: GenServer.call(OMG.Watcher.ExitProcessor, {:challenge_exits, []}, 5000)
watcher-0 watcher     ** (EXIT) an exception was raised:
watcher-0 watcher         ** (Ecto.ConstraintError) constraint error when attempting to insert struct:
watcher-0 watcher
watcher-0 watcher     * unique: ethevents_pkey
watcher-0 watcher
watcher-0 watcher If you would like to convert this constraint into an error, please
watcher-0 watcher call unique_constraint/3 in your changeset and define the proper
watcher-0 watcher constraint name. The changeset has not defined any constraint.
watcher-0 watcher
watcher-0 watcher             (ecto) lib/ecto/repo/schema.ex:574: anonymous fn/4 in Ecto.Repo.Schema.constraints_to_errors/3
watcher-0 watcher             (elixir) lib/enum.ex:1327: Enum."-map/2-lists^map/1-0-"/2
watcher-0 watcher             (ecto) lib/ecto/repo/schema.ex:559: Ecto.Repo.Schema.constraints_to_errors/3
watcher-0 watcher             (ecto) lib/ecto/repo/schema.ex:222: anonymous fn/14 in Ecto.Repo.Schema.do_insert/4
watcher-0 watcher             (ecto) lib/ecto/repo/schema.ex:774: anonymous fn/3 in Ecto.Repo.Schema.wrap_in_transaction/6
watcher-0 watcher             (ecto) lib/ecto/adapters/sql.ex:576: anonymous fn/3 in Ecto.Adapters.SQL.do_transaction/3
watcher-0 watcher             (db_connection) lib/db_connection.ex:1283: DBConnection.transaction_run/4
watcher-0 watcher             (db_connection) lib/db_connection.ex:1207: DBConnection.run_begin/3
watcher-0 watcher     (elixir) lib/gen_server.ex:989: GenServer.call/3
watcher-0 watcher     lib/ethereum_event_listener.ex:101: OMG.API.EthereumEventListener.sync_height/2
watcher-0 watcher     lib/ethereum_event_listener.ex:84: OMG.API.EthereumEventListener.handle_info/2
watcher-0 watcher     (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
watcher-0 watcher     (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
watcher-0 watcher     (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
watcher-0 watcher Last message: :sync⋅
watcher-0 watcher 2019-02-18 04:52:56.807 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for in_flight_exit_processor⋅
watcher-0 watcher 2019-02-18 04:52:56.809 [info] module=OMG.API.EthereumEventListener function=init/1 ⋅Starting EthereumEventListener for exit_processor⋅
watcher-0 watcher 2019-02-18 04:52:56.812 [info] module=application_controller function=info_exited/3 ⋅Application omg_watcher exited: shutdown⋅
watcher-0 watcher 2019-02-18 04:52:56.874 [error] module=gen_event function=report_error/5 ⋅:gen_event handler Sentry.Logger installed in :error_logger terminating
watcher-0 watcher ** (stop) exited in: GenServer.call(Sentry.TaskSupervisor, {:start_task, [{:nonode@nohost, :error_logger, #PID<0.356.0>}, [#PID<0.356.0>], :monitor, {:erlang, :apply, [#Function<3.89165438/0 in Sentry.Client.do_send_event/3>, []]}], :temporary, nil}, :infinity)
watcher-0 watcher     ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
watcher-0 watcher Last message: {:error_report, #PID<0.393.0>, {#PID<0.419.0>, :crash_report, [[initial_call: {OMG.Watcher.ExitProcessor, :init, [:Argument__1]}, pid: #PID<0.419.0>, registered_name: OMG.Watcher.ExitProcessor, error_info: {:error, %Ecto.ConstraintError{constraint: "ethevents_pkey", message: "constraint error when attempting to insert struct:\n\n    * unique: ethevents_pkey\n\nIf you would like to convert this constraint into an error, please\ncall unique_constraint/3 in your changeset and define the proper\nconstraint name. The changeset has not defined any constraint.\n", type: :unique}, [{Ecto.Repo.Schema, :"-constraints_to_errors/3-fun-1-", 4, [file: 'lib/ecto/repo/schema.ex', line: 574]}, {Enum, :"-map/2-lists^map/1-0-", 2, [file: 'lib/enum.ex', line: 1327]}, {Ecto.Repo.Schema, :constraints_to_errors, 3, [file: 'lib/ecto/repo/schema.ex', line: 559]}, {Ecto.Repo.Schema, :"-do_insert/4-fun-1-", 14, [file: 'lib/ecto/repo/schema.ex', line: 222]}, {Ecto.Repo.Schema, :"-wrap_in_transaction/6-fun-0-", 3, [file: 'lib/ecto/repo/schema.ex', line: 774]}, {Ecto.Adapters.SQL, :"-do_transaction/3-fun-1-", 3, [file: 'lib/ecto/adapters/sql.ex', line: 576]}, {DBConnection, :transaction_run, 4, [file: 'lib/db_connection.ex', line: 1283]}, {DBConnection, :run_begin, 3, [file: 'lib/db_connection.ex', line: 1207]}]}, ancestors: [OMG.Watcher.Supervisor, OMG.Watcher.RootSupervisor, #PID<0.394.0>], message_queue_len: 5, messages: [{:"$gen_call", {#PID<0.424.0>, #Reference<0.210763670.3553624067.91748>}, {:piggyback_exits, []}}, {:"$gen_call", {#PID<0.423.0>, #Reference<0.210763670.3553624065.97256>}, {:new_in_flight_exits, []}}, {:"$gen_call", {#PID<0.422.0>, #Reference<0.210763670.3553624066.91237>}, {:challenge_exits, []}}, {:"$gen_call", {#PID<0.566.0>, #Reference<0.210763670.3553624066.91528>}, :check_validity}, {:"$gen_call", {#PID<0.557.0>, #Reference<0.210763670.3553624065.97451>}, :check_validity}], links: [#PID<0.396.0>], dictionary: [], trap_exit: false, status: :running, heap_size: 6772, stack_size: 27, reductions: 66248], []]}}⋅
```

**Comment by pdobacz** [comment]
@jbunce did you start syncing from scratch in `:prod` or just switch the modes, having done anything in `:dev` earlier.

I think one even would have to do the `mix ecto.reset` in `:prod` already, as it sets up the Postgres databases properly - did you?

I'd expect such error to pop up, when some data is being added twice to Postgres

**Comment by jbunce** [comment]
Yes, `mix ecto.reset` was followed for this deployment. It was a fresh sync.

**Comment by pnowosie** [comment]
Events table is tricky as we [generate record's Pk](https://github.com/omisego/elixir-omg/blob/c652f1384105e7875c48b95130807d9cd218129d/apps/omg_watcher/lib/db/eth_event.ex#L96) from event data. There wasn't good candidate for unique identified in Ethereum at the time. 
However key generation is unrelated to the application environment.

**Comment by pdobacz** [comment]
I checked that on e75bef9d953cbb5bc820ac9a333050e35c0973c7 and 6cdb10c74d1319ff799c26df85bd6d711bed0998, both work for me. Did like this:

```
$ rm -rf ~/.omg/data_staging && mix ecto.reset --no-start && mix run --no-start -e 'OMG.DB.init' --config ~/config_staging.exs

...
$ MIX_ENV=prod DATABASE_URL=postgres://omisego_dev:omisego_dev@localhost/omisego_dev mix xomg.watcher.start --convenience --config ~/config_staging.exs

... syncs, syncs, done!
2019-02-21 16:14:23.624 [info] module=OMG.Watcher.BlockGetter.Core function=apply_block/2 ⋅Applied block: #76000, from eth height: 3907063, eth height done?: true⋅
2019-02-21 16:14:27.250 [info] module=OMG.Watcher.ExitProcessor function=handle_call/3 ⋅Recognized exits: [%{amount: 1000, currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>, eth_height: 3902925, owner: <<131, 22, 38, 25, 106, 35, 178, 202, 36, 239, 138, 121, 90, 96, 132, 55, 44, 185, 161, 193>>, utxo_pos: 56002000000000}, %{amount: 900, currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>, eth_height: 3903090, owner: <<131, 22, 38, 25, 106, 35, 178, 202, 36, 239, 138, 121, 90, 96, 132, 55, 44, 185, 161, 193>>, utxo_pos: 59000000480000}, %{amount: 901, currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>, eth_height: 3903108, owner: <<211, 135, 183, 91, 96, 210, 71, 152, 137, 144, 134, 120, 136, 127, 42, 235, 50, 221, 215, 8>>, utxo_pos: 59000000470000}]⋅
2019-02-21 16:14:33.252 [info] module=OMG.Watcher.ExitProcessor function=handle_call/3 ⋅Recognized challenges: [%{eth_height: 3903129, utxo_pos: 59000000470000}]⋅
```

It either is bad luck and running into https://github.com/omisego/elixir-omg/issues/454 or writing into a populated PostgreSQL database combined with syncing from scratch (i.e. LevelDB wiped clean).

Can you please double check @jbunce that the `mix ecto.reset` did its thing on the correct DB? Not sure how the database is configured but maybe one needs to `MIX_ENV=prod mix ecto.reset`?

**Comment by jbunce** [comment]
tl;dr: Don't run the Watcher in `MIX_ENV=prod` after it has state from running in `MIX_ENV=dev`. Ari is running a prod Watcher now

**Comment by pdobacz** [comment]
> tl;dr: Don't run the Watcher in `MIX_ENV=prod` after it has state from running in `MIX_ENV=dev`. Ari is running a prod Watcher now

For completness sake: the tl;dr is on point, but the gist of the problem is likely just that `prod` and `dev` use a different postgres database instance, so generally one can switch `MIX_ENV` in flight with some careful configs/migrating data or sth similar.

Thanks for revisiting!
